File: Javascript/string-compression.js
// This algorithm is to compress a string, to give a count of equivalent consecutive 
// letters, such that "aabbccddeeeeeeea" would equal "2a2b2c2d7e1a". If the compressed
// string has a length equal to or greater than the original string, the original string
// will be returned.

// Write the string compression function

var compressString = function(str) {
  newStr = "";
  counter = 1;
  currentLetter = str[0];
  for (var i = 1; i < str.length; i++) {
    if (str[i] === currentLetter) {
      counter += 1;
    } else {
      newStr += currentLetter + counter.toString();
      counter = 1;
      currentLetter = str[i];
    }
  }
  newStr += currentLetter + counter.toString();
  if (newStr.length < str.length) {
    return newStr;
  } else {
    return str;
  }
}

// Create strings and call string compression function on them

strOne = "aabbccddeeeeeeea";

console.log(compressString(strOne));

strTwo = "aabcdef";

console.log(compressString(strTwo));

File: Ruby/string-compression.rb
# This algorithm is to compress a string, to give a count of equivalent consecutive
# letters, such that "aabbcddeeeeaa" would equal "2a2b1c2d4e2a". If the compressed
# string has a length equal to or greater than the original string, the original string
# will be returned.

# Write the string compression method

def string_compression(str)
  new_str = ""
  counter = 1
  current_letter = str[0]
  str.split("")[1..-1].each do |letter|
    if letter == current_letter
      counter += 1
    else
      new_str += counter.to_s + current_letter
      current_letter = letter
      counter = 1
    end
  end
  new_str += counter.to_s + current_letter
  if new_str.length >= str.length
    return str
  else
    return new_str
  end
end

# Create strings and call string compression method on them

str = "aaabbeeddccccdd"

puts string_compression(str)

str = "abcdef"

puts string_compression(str)